packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```
@@ -0,0 +1,573 @@
+
+    const logMeta = {persistentKey: toHexString(persistentKey)};
+
+    // reload from disk or db based on closest checkpoint
+    this.logger.verbose("Reload: read state", logMeta);
+    const newStateBytes = await this.persistentApis.read(persistentKey);
+    if (newStateBytes === null) {
+      this.logger.verbose("Reload: read state failed", logMeta);
+      return null;
+    }
+    this.logger.verbose("Reload: read state successfully", logMeta);
+    this.metrics?.stateRemoveCount.inc({reason: RemovePersistedStateReason.reload});
+    this.metrics?.stateReloadSecFromSlot.observe(this.clock?.secFromSlot(this.clock?.currentSlot ?? 0) ?? 0);
+    const closestState = this.findSeedStateToReload(cp) ?? this.getHeadState?.();
+    if (closestState == null) {
+      throw new Error("No closest state found for cp " + toCheckpointKey(cp));
+    }
+    this.metrics?.stateReloadEpochDiff.observe(Math.abs(closestState.epochCtx.epoch - cp.epoch));
+    this.logger.verbose("Reload: found closest state", {...logMeta, seedSlot: closestState.slot});
+    const timer = this.metrics?.stateReloadDuration.startTimer();
+
+    try {
+      const newCachedState = loadCachedBeaconState(closestState, newStateBytes, {
+        shufflingGetter: this.shufflingCache.getSync.bind(this.shufflingCache),
+      });
+      timer?.();
+      this.logger.verbose("Reload state successfully", {
+        ...logMeta,
+        stateSlot: newCachedState.slot,
+        seedSlot: closestState.slot,
+      });
+      // only remove persisted state once we reload successfully
+      this.persistentApis
+        .remove(persistentKey)
+        .catch((e) => this.logger.debug("Error removing persisted state", logMeta, e));
+      this.cache.set(cpKey, newCachedState);
+      // don't prune from memory here, call it at the last 1/3 of slot 0 of an epoch
+      return newCachedState;
+    } catch (e) {
+      this.logger.debug("Error reloading state from disk", logMeta, e as Error);
+      return null;
```

null here can mean many things at once, why not propagate the error up?